### PR TITLE
use bufio.Writer instead of binary.Write to reduce memory footprint while serializing

### PR DESCRIPTION
**Pull request by thanhpk**
While writing (or reading) a very big bitset (few GBs) into a file, the memory usage will be doubled, crashing the program.
This is because you are converting the whole set into another big `[]byte` slice. (through `binary.Write`)
This pull request fixs the issue by not trying to convert the whole set (`[]uint64`) into `[]byte`, but convert one by one item and push it to a buffered stream.
The memory usage is the same after calling `WriteTo` or `ReadFrom`

Refs:
`bufio` is the recomented way to read and writing buffered data. (https://pkg.go.dev/bufio#pkg-overview)
A loop while reading out from `bufio.Reader` is also quite common (https://pkg.go.dev/bufio#example-Scanner.Bytes)

### CODE REVIEW

**Comment by thanhpk** [comment]
In the implementation of binary.Write, its clear that binary.Write will allocates another []byte with same size with the input.
https://github.com/golang/go/blob/66865363f017a8d4cb0b07d84a3a6117fcf1cd30/src/encoding/binary/binary.go#L341

I ran a benchmark to see how much binary.Write allocated compare to buffio.Write
```
$ go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: github.com/thanhpk/testbitset
cpu: Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz
BenchmarkBinaryWrite-12    	       1	2727863238 ns/op	3200014368 B/op	      12 allocs/op
BenchmarkBufioWriter-12    	       1	3783645077 ns/op	1600008360 B/op	       9 allocs/op
PASS
ok  	github.com/thanhpk/testbitset	6.922s
```
As you can see, the binary version is faster but use 2x memory. Note that this is only true for large data.
Here is the code for the benchmark
```
package main

import (
	"bufio"
	"encoding/binary"
	"os"
	"testing"
)

func BenchmarkBinaryWrite(b *testing.B) {
	f, err := os.OpenFile("./binary.dat", os.O_WRONLY|os.O_CREATE, 0600)
	checkErr(err)
	writer := bufio.NewWriter(f)

	data := make([]uint64, 200000000) // 1.5GB

	for i := 0; i < b.N; i++ {
		err = binary.Write(writer, binary.LittleEndian, data)
		checkErr(err)
	}

	err = writer.Flush()
	checkErr(err)

	err = os.Remove("./binary.dat")
	checkErr(err)
}

func BenchmarkBufioWriter(b *testing.B) {
	f, err := os.OpenFile("./buffio.dat", os.O_WRONLY|os.O_CREATE, 0600)
	checkErr(err)

	writer := bufio.NewWriter(f)

	data := make([]uint64, 200000000) // 1.5GB

	for i := 0; i < b.N; i++ {
		var item = make([]byte, 8) // for serializing uint64
		for _, x := range data {
			binary.LittleEndian.PutUint64(item, x)
			_, err := writer.Write(item)
			checkErr(err)
		}
	}

	err = writer.Flush()
	checkErr(err)

	err = os.Remove("./buffio.dat")
	checkErr(err)
}

func checkErr(err error) {
	if err != nil {
		panic(err)
	}
}
```

**Comment by lemire** [comment]
Thanks.

@josharian : can you comment?

File: bitset.go [hunk 1]
```diff
@@ -895,8 +895,22 @@ func (b *BitSet) WriteTo(stream io.Writer) (int64, error) {
 	}
 
 	// Write set
-	err = binary.Write(stream, binaryOrder, b.set)
-	return int64(b.BinaryStorageSize()), err
+	// current implementation of bufio.Writer is more memory efficient than
+	// binary.Write for large set
+	writer := bufio.NewWriter(stream)
+	var item = make([]byte, 8)     // for serializing uint64
+	var n = binary.Size(uint64(0)) // number of bytes written
```
Comment (lemire) [inline bitset.go:902]: That seems odd, doesn't it? You assume that the size is 8, and then you next compute it? Do I get this right?

File: bitset.go [hunk 2]
```diff
@@ -895,8 +895,22 @@ func (b *BitSet) WriteTo(stream io.Writer) (int64, error) {
...
+	for _, x := range b.set {
+		binaryOrder.PutUint64(item, x)
+		nn, err := writer.Write(item)
+		if err != nil {
+			return int64(n+nn), err
+		}
+		n += nn
```
Comment (lemire) [inline bitset.go:909]: Why would you do all of these additions?

File: bitset.go [hunk 3]
```diff
@@ -915,9 +929,20 @@ func (b *BitSet) ReadFrom(stream io.Reader) (int64, error) {
...
 	// Read remaining bytes as set
-	err = binary.Read(stream, binaryOrder, newset.set)
-	if err != nil {
-		return 0, err
+	// current implementation bufio.Reader is more memory efficient than
+	// binary.Read for large set
+	reader := bufio.NewReader(stream)
+	i := 0
+	var item = make([]byte, 8) // one uint64
+	for {
```
Comment (lemire) [inline bitset.go:937]: Why not go with the standard...

```
for i:=0 ....
```

**Comment by lemire** [comment]
@thanhpk Let us push this forward. Please have a look at my comments.


@josharian 

> Yes, binary.Write allocates the entire buffer up front. This is arguably a bug: It could do internal batching for large slices. I'm not sure how much excitement there'd be upstream for fixing it.

It is disappointing code.

**Comment by josharian** [comment]
I suspect that the upstream response is: "binary.Write is not meant for giant data structures; for very large things you probably want enough control that you should write it yourself". But don't take my word for it: File an issue. I don't recall having seen anything specifically about this; the closest I recall are https://github.com/golang/go/issues/27757 and https://github.com/golang/go/issues/27403.

**Comment by thanhpk** [comment]
@lemire I fixed the issues you commented.

**Comment by lemire** [comment]
Merged.

**Pull request merged**